File: app/services/execution/execution_service.py
from uuid import uuid4
from decimal import Decimal

from app.models.order import Order
from app.models.trade import Trade

from app.repositories.order_repository import (
    OrderRepository
)

from app.repositories.trade_repository import (
    TradeRepository
)


class ExecutionService:

    def __init__(
        self,
        db
    ):

        self.order_repo = OrderRepository(db)

        self.trade_repo = TradeRepository(db)

    async def persist_execution(
        self,
        execution_result: dict
    ):

        order = Order(
            id=str(uuid4()),
            symbol=execution_result["symbol"],
            side=execution_result["side"],
            order_type="MARKET",
            quantity=Decimal(
                str(execution_result["quantity"])
            ),
            price=Decimal(
                str(execution_result["price"])
            ),
            status="FILLED"
        )

        saved_order = await self.order_repo.create(
            order
        )

        trade = Trade(
            id=str(uuid4()),
            order_id=saved_order.id,
            symbol=execution_result["symbol"],
            side=execution_result["side"],
            quantity=Decimal(
                str(execution_result["quantity"])
            ),
            price=Decimal(
                str(execution_result["price"])
            )
        )

        await self.trade_repo.create(
            trade
        )

        return saved_order

File: backend/src/services/executionService.ts
import { randomUUID } from "crypto";
import { OrderRepository } from "../repositories/orderRepository";
import { TradeRepository } from "../repositories/tradeRepository";
import { PositionService } from "./positionService";
import { AuditService } from "./auditService";

export class ExecutionService {
  static async persistExecution(executionResult: {
    symbol: string;
    side: string;
    quantity: number;
    price: number;
  }) {
    const orderId = randomUUID();
    const order = await OrderRepository.create({
      id: orderId,
      symbol: executionResult.symbol,
      side: executionResult.side,
      order_type: "MARKET",
      quantity: executionResult.quantity,
      price: executionResult.price,
      status: "FILLED",
    });

    const tradeId = randomUUID();
    await TradeRepository.create({
      id: tradeId,
      order_id: order.id,
      symbol: executionResult.symbol,
      side: executionResult.side,
      quantity: executionResult.quantity,
      price: executionResult.price,
    });

    const side = executionResult.side;
    const symbol = executionResult.symbol;
    const price = executionResult.price;
    const quantity = executionResult.quantity;

    if (side === "BUY") {
      await PositionService.createPosition(
        symbol,
        quantity,
        price,
        price * 0.98,
        price * 1.03
      );
      await AuditService.logEvent(
        "POSITION",
        "POSITION_OPENED",
        symbol,
        { side, quantity, price }
      );
    } else if (side === "SELL") {
      await PositionService.closePosition(symbol);
      await AuditService.logEvent(
        "POSITION",
        "POSITION_CLOSED",
        symbol,
        { side, quantity, price }
      );
    }

    await AuditService.logEvent(
      "EXECUTION",
      "ORDER_FILLED",
      order.id,
      executionResult
    );

    return order;
  }
}
export default ExecutionService;
